### PR TITLE
fix: handle relative assets paths in server `fetch` correctly

### LINKED ISSUE #12081: Can't fetch assets using Kits `fetch` and relative path
### Describe the bug

According to the documentation ([here](https://kit.svelte.dev/docs/load#making-fetch-requests) and [here](https://kit.svelte.dev/docs/load#making-fetch-requests)), kits `fetch` should be able to handle relative paths on the server.

This works in development, but not in production (I've tested Cloudflare and node adapters).

If I do this:

```js
// +server.ts
import asset from "$lib/image.png";

export const GET = async ({ fetch: sk_fetch }) => {
	const _asset = await sk_fetch(asset);

	// we always end up here
	if (!_asset.ok)  return error(404, "Not found");
	
	return _asset;
};
```

The response is always this exact 404 (as far as I can tell): 
https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/kit/src/runtime/server/respond.js#L118-L120

We end up here, because the `server_fetch` doesn't recognise this path as an asset.
Following check seams to be the problem:

https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/kit/src/runtime/server/fetch.js#L84-L87

as it doesn't check `manifest._.server_assets`

---

I've created an exhaustive reproduction that shows differences between sk/global fetch and absolute/relative path here:
https://github.com/mastermakrela/sveltekit-relative-assets-fetch-bug

---

### Solutions considered

1. Use [`read`](https://kit.svelte.dev/docs/modules#$app-server-read) 
    - works only for node
    - currently can't be implemented for Cloudflare, because `read` is sync and Cloudflares `env.ASSETS.fetch` isn't 

2. Change [`is_asset`](https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/kit/src/runtime/server/fetch.js#L84) check in Kit to
    ```js
    const is_asset =
    	manifest.assets.has(filename) ||
    	Object.prototype.hasOwnProperty.call(manifest._.server_assets, filename);
    ```

3. Change [`is_static_asset`](https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/adapter-cloudflare/src/worker.js#L33-L34) check in Cloudflare adapter to:
    ```js
    is_static_asset =
    	manifest.assets.has(filename) ||
    	manifest.assets.has(filename + '/index.html') ||
    	Object.prototype.hasOwnProperty.call(manifest._.server_assets, filename);
    ```

4. Use global fetch
    - as shown in https://github.com/mastermakrela/sveltekit-relative-assets-fetch-bug the simplest solution is just to use global fetch and absolute path
    - this isn't optimal because for each asset a whole request is needed (performance cost)
    - and one has to construct the absolute URL from vites asset import

---

Semantically async `read` + Cloudflare implementation, would be most satisfying, but it's probably leas viable, because of breaking change.

But independent of read, **not being able to fetch assets using relative path on server** seams to be a bug, that should be fixed (or if not at least a warning in dev mode would be nice).

### Reproduction

https://github.com/mastermakrela/sveltekit-relative-assets-fetch-bug

### Logs

_No response_

### System Info

```Shell
System:
    OS: macOS 14.4.1
    CPU: (8) arm64 Apple M1
    Memory: 69.52 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.7.1 - /opt/homebrew/bin/node
    npm: 10.5.0 - /opt/homebrew/bin/npm
    pnpm: 8.15.6 - /opt/homebrew/bin/pnpm
    bun: 1.0.25 - ~/.bun/bin/bun
  Browsers:
    Brave Browser: 123.1.64.109
    Chrome: 123.0.6312.87
    Edge: 112.0.1722.54
    Safari: 17.4.1
  npmPackages:
    @sveltejs/adapter-auto: ^3.2.0 => 3.2.0 
    @sveltejs/adapter-cloudflare: ^4.2.0 => 4.2.0 
    @sveltejs/adapter-node: ^5.0.1 => 5.0.1 
    @sveltejs/vite-plugin-svelte: ^3.0.2 => 3.0.2 
    svelte: 5.0.0-next.90 => 5.0.0-next.90 
    vite: ^5.2.7 => 5.2.7
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by mastermakrela**
## The Problem

Fixes #12081 (and _maybe_ helps with #3850).

TL;DR:
```js
// +server.ts
import asset from "$lib/image.png";

export const GET = async ({ fetch }) => { return fetch(asset); };
```

works in dev and preview, but always returns [`404`](https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/kit/src/runtime/server/respond.js#L118-L120) with adapter-cloudflare and adapter-node.

More details and exhaustive reproduction in #12081 and [here](https://github.com/mastermakrela/sveltekit-relative-assets-fetch-bug)

## The Solution

I've extended all places that check `manifest.assets` to also check `manifest._.server_assets`. 
This has two benefits:

1. the adapters can return the file directly without calling the SvelteKit server
2. `server_fetch` (in `packages/kit/src/runtime/server/fetch.js`) can return the file earlier (previously it fell through to [`response`](https://github.com/sveltejs/kit/blob/df6641a701f8155d7023ab6a5b17ef39df31d730/packages/kit/src/runtime/server/fetch.js#L127))

This should always work because since #11649 all server assets should be copied to the output folder ([source](https://github.com/sveltejs/kit/issues/10594#issuecomment-1901430103)).

## Other improvements

If [`read`](https://kit.svelte.dev/docs/modules#$app-server-read) implementation is available, `server_fetch` will use it to get the files directly — should be theoretically faster than a fetch.

This PR is similar to what #3850 proposed, but because `state.read`/`options.read` expects a function that returns a buffer and `read_implementation` returns a stream, they aren't compatible. Also, it still wouldn't work with Cloudflare, because there the “file read” is an async fetch.

But this PR brings us as close as possible to #3850 as currently possible, because:
- node will use [`read`](https://kit.svelte.dev/docs/modules#$app-server-read) — skips overhead of fetch
- Cloudflare will fetch static asset directly

and both of them (and maybe more adapters that I didn't test) will work with relative paths on the server.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

File: .changeset/dirty-rivers-tickle.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"@sveltejs/adapter-cloudflare-workers": minor
+"@sveltejs/adapter-cloudflare": minor
+"@sveltejs/adapter-netlify": minor
+"@sveltejs/kit": minor
```
Comment (eltigerchino) [inline .changeset/dirty-rivers-tickle.md:5]:
```suggestion
"@sveltejs/adapter-cloudflare-workers": patch
"@sveltejs/adapter-cloudflare": patch
"@sveltejs/adapter-netlify": patch
"@sveltejs/kit": patch
```

File: packages/adapter-cloudflare-workers/files/entry.js [hunk 2]
```diff
@@ -57,7 +57,10 @@ export default {
 		const filename = stripped_pathname.substring(1);
 		if (filename) {
 			is_static_asset =
-				manifest.assets.has(filename) || manifest.assets.has(filename + '/index.html');
+				manifest.assets.has(filename) ||
+				manifest.assets.has(filename + '/index.html') ||
+				Object.prototype.hasOwnProperty.call(manifest._.server_assets, filename) ||
+				Object.prototype.hasOwnProperty.call(manifest._.server_assets, filename + '/index.html');
```
Comment (eltigerchino) [inline packages/adapter-cloudflare-workers/files/entry.js:63]:
```suggestion
				filename in manifest._.server_assets ||
				filename + '/index.html' in manifest._.server_assets);
```

File: packages/kit/test/apps/basics/src/routes/endpoint-output/fetch-asset/absolute/+server.js
Comment (eltigerchino) [inline packages/kit/test/apps/basics/src/routes/endpoint-output/fetch-asset/absolute/+server.js:1]: Could we change the test to import a text file instead? We can ensure it doesn't get inlined by using the Vite `?url` query parameter in the import path.

File: .changeset/dirty-rivers-tickle.md (hunk 1 above)
Comment (mastermakrela) [reply .changeset/dirty-rivers-tickle.md]: Thank you!
Sorry, it's my first contribution, so I wasn't sure and the change seamed _minor_ ;)

File: packages/adapter-cloudflare-workers/files/entry.js (hunk 2 above)
Comment (mastermakrela) [reply packages/adapter-cloudflare-workers/files/entry.js]: Because I've written it almost half a year ago, I can't recall why I used `hasOwnProperty` 😅
I have vague recollection that I tried `in` and something didn't work, but I tested it again today, and it seamed to work as intended, so I've changed it (in all 3 adapters: Cloudflare, Cloudflare-workers and netlify).

File: packages/kit/test/apps/basics/src/routes/endpoint-output/fetch-asset/absolute/+server.js
Comment (mastermakrela) [reply packages/kit/test/apps/basics/src/routes/endpoint-output/fetch-asset/absolute/+server.js:1]: Of course, I've changed it to a text file. :)

**Comment by mastermakrela** [comment]
@eltigerchino thank you for the review :)

I've corrected all the issues you've noted and tested the Cloudflare adapter again.

I see some tests in CI have failed, but as far as I understand, two of them should fail (Vercel ones) 
and the third one (`/prerendering/mutative-endpoint`) does not seam to be connected with my changes.

Is there anything else I could/should do?

File: .changeset/dirty-rivers-tickle.md (hunk 1 above)
Comment (eltigerchino) [reply .changeset/dirty-rivers-tickle.md]: No worries! Here's where the concept is taken from: https://semver.org/

File: .changeset/dirty-rivers-tickle.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+---
+"@sveltejs/adapter-cloudflare-workers": patch
+"@sveltejs/adapter-cloudflare": patch
+"@sveltejs/adapter-netlify": patch
+"@sveltejs/kit": patch
+---
+
+fix: fix handling of relative paths in fetch on the server
```
Comment (eltigerchino) [inline .changeset/dirty-rivers-tickle.md:8]:
```suggestion
fix: correctly handle relative paths when fetching assets on the server
```

**Review by eltigerchino — APPROVED** [review]

**Comment by eltigerchino** [comment]
Thank you! This is a really good contribution. Don't worry about some of the tests which are flaky and pass when re-run.

**Review by dummdidumm — APPROVED** [review]
Fantastic work, and great PR description, was really easy to review because of that 👍
